File: leti/3304/sandin/2/src/main/java/org/lab2/sandin/LockFreePriorityQueue.java
package org.lab2.sandin;

import java.util.AbstractQueue;
import java.util.Iterator;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicMarkableReference;

public class LockFreePriorityQueue<E extends Comparable<E>> extends AbstractQueue<E> implements PriorityQueue<E>{

    private AtomicMarkableReference<Node<E>> head;
    private AtomicMarkableReference<Node<E>> tail;
    private AtomicInteger numOfNodes;

    LockFreePriorityQueue(){
        Node<E> nodeTail = new Node<>(null);
        Node<E> nodeHead = new Node<>(null,nodeTail);
        head = new AtomicMarkableReference<>(nodeHead, false);
        tail = new AtomicMarkableReference<>(nodeTail, false);
        numOfNodes = new AtomicInteger(0);
    }
    
    @Override
    public Iterator<E> iterator() {
        return null;
    }

    @Override
    public boolean offer(E e) {
        Node<E> newNode = new Node<>(e);
        
        while (true)
        {

            Node<E> prev = head.getReference();
            Node<E> current = prev.getNext();
            Node<E> next = current.getNext();
            
            while (true)
            {
                if (next == null || e.compareTo(current.value) < 0)
                {
                    break;
                }
                else if (current.isMarked())
                {
                    if (prev.next.compareAndSet(current, next, false, false))
                    {
                        numOfNodes.decrementAndGet();
                    }
                }
                
                prev = current;
                current = current.getNext();
                next = current.getNext();
            }
            
            newNode.next.set(current, false);
            if (prev.next.compareAndSet(current, newNode, false, false))
            {
                numOfNodes.incrementAndGet();
                return true;
            }         
        }
    }

    @Override
    public E poll() {
        Node<E> first = null;
        Node<E> current = null;
        Node<E> next = null;
        Node<E> last = null;
        
        while (true)
        {
            first = head.getReference();
            current = first.getNext();
            next = current.getNext();
            last = tail.getReference();
            
            if (current == last)
            {
                return null;
            }
            
            if (current.isMarked())
	            {
	                if (first.next.compareAndSet(current, next, false, false))
	                {
	                    numOfNodes.decrementAndGet();
	                }
	            }
	            else
	            {
	                E data = current.value;
	                if (current.next.compareAndSet(next, next, false, true))
	                {
	                    if (first.next.compareAndSet(current, next, false, false))
	                    {
	                        numOfNodes.decrementAndGet();
	                    }
	                    
	                    return data;
	                }
	            }
            }
    }

    @Override
    public E peek() {
        Node<E> first;
        Node<E> current;
        Node<E> next;
        Node<E> last;
        
        while (true)
        {
            first = head.getReference();
            current = first.getNext();
            next = current.getNext();
            last = tail.getReference();

            if (current.isMarked())
            {
                if (first.next.compareAndSet(current, next, false, false))
                {
                    numOfNodes.decrementAndGet();
                }
            }
            else if (current != last)
            {
                return current.value;
            }
            else 
            {
                return null;
            }
        }
    }

    @Override
    public  boolean isEmpty(){
        Node<E> first = null;
        Node<E> current = null;
        Node<E> last = null;

        while(true)
        {
            first = head.getReference();
            current = first.getNext();
            last = tail.getReference();
            
            if (current == last)
            {
                return true;
            } 

            if (current.isMarked())
            {
                if (first.next.compareAndSet(current, current.getNext(), false, false))
                {
                    numOfNodes.decrementAndGet();
                }
                
                continue;
            }
  
            return false;
        }
    }
    
    @Override
    public int size() {
        return numOfNodes.get();
    }
    
    protected class Node <E extends Comparable<E>> {
        public E value;
        public AtomicMarkableReference<Node<E>> next;

        Node(E value){
            this.value = value;
            this.next = new AtomicMarkableReference<>(null, false);
        }

        Node (E value, Node<E> next){
            this.value = value;
            this.next = new AtomicMarkableReference<>(next, false);
        }
        
        boolean isMarked() {
            return next.isMarked();
        }

        Node<E> getNext() {
            return next.getReference();
        }
    }
}

File: itmo/M4236/kuznetsov/hw2/src/main/kotlin/ru/itmo/hpc/LockFreePriorityQueue.kt
package ru.itmo.hpc

import java.util.*
import java.util.concurrent.atomic.AtomicMarkableReference

/**
 * Implementation based on the concepts of Harris list
 * @link https://timharris.uk/papers/2001-disc.pdf
 */
class LockFreePriorityQueue<E: Comparable<E>>: AbstractQueue<E>(), PriorityQueue<E> {

    override fun offer(e: E): Boolean {

        while (true) {
            val (predecessor, current) = getInsertPosition(e)

            if (predecessor.compareAndSetNext(
                            current,
                            Node(e, current),
                            false,
                            false)) {
                return true
            }
        }
    }

    override fun iterator(): MutableIterator<E> {
        TODO("not implemented")
    }

    override val size: Int
        get() = TODO("not implemented") //To change initializer of created properties use File | Settings | File Templates.

    override fun peek(): E? {
        return getInsertPosition().current.data
    }

    /**
     * It does not physically remove any node from queue,
     * but marks node that follows *after* the first present element.
     *
     * Therefore, set mark means that *previous* node must not be considered
     * in further operations.
     */
    override fun poll(): E? {
        while (true) {
            val appropriateNode = getInsertPosition().current
            if (appropriateNode == TAIL)
                return null
            val next = appropriateNode.nextNode
            if (appropriateNode.compareAndSetNext(next, next, false, true)) {
                return appropriateNode.data
            }
        }
    }

    override fun isEmpty(): Boolean {
        val (prev, curr) = getInsertPosition()
        return (prev == HEAD && curr == TAIL)
    }

    inner class Node<E: Comparable<E>>(
            val data: E? = null, // Only for HEAD and TAIL
            val next: Node<E>? = null // Only for TAIL
    ): Comparable<Node<E>> {

        private val nextAtomicReference: AtomicMarkableReference<Node<E>?> = AtomicMarkableReference(next, false)

        val nextNode get() = nextAtomicReference.reference!!

        val isNextMarked get() = nextAtomicReference.isMarked

        override fun compareTo(other: Node<E>): Int =
                when (other) {
                    HEAD -> 1
                    TAIL -> -1
                    else -> data!!.compareTo(other.data!!)
                }

        operator fun compareTo(other: E): Int =
                when (other) {
                    HEAD -> 1
                    TAIL -> -1
                    else -> data!!.compareTo(other)
                }

        fun compareAndSetNext(
                expectedReference: Node<E>,
                newReference: Node<E>,
                expectedMark: Boolean = false,
                newMark: Boolean = false
        ) = nextAtomicReference.compareAndSet(expectedReference, newReference, expectedMark, newMark)
    }

    data class SequentialPair<E: Comparable<E>>(
            val predecessor: LockFreePriorityQueue<E>.Node<E>,
            val current: LockFreePriorityQueue<E>.Node<E>
    )

    /**
     * Get pair of consequent nodes to insert new element right between them
     */
    private fun getInsertPosition(e: E? = null): SequentialPair<E> {
        var previousNode = HEAD
        var currentNode = HEAD.nextNode
        var nextNode: Node<E>

        while (true) {
            var temporalNode = HEAD
            var temporalNextNode = HEAD.nextNode

            do {
                if (!temporalNode.isNextMarked) {
                    previousNode = temporalNode
                    currentNode = temporalNextNode
                }
                temporalNode = temporalNode.nextNode
                if (temporalNode === TAIL) {
                    break
                }
                temporalNextNode = temporalNode.nextNode
            } while (temporalNode.isNextMarked || (e != null && temporalNode <= e))

            nextNode = temporalNode

            if (previousNode.compareAndSetNext(currentNode, nextNode, false, false)) {
                if (nextNode != TAIL && nextNode.isNextMarked) {
                    continue
                } else {
                    return SequentialPair(previousNode, nextNode)
                }
            }
        }
    }

    private val TAIL: Node<E> = Node()
    private val HEAD: Node<E> = Node(next = TAIL)
}